tools/src/scie_pants/pants_version.py

```diff
@@ -206,3 +175,72 @@ def configure_version():
     return configure_version, determine_tag_version(
         ptex, pants_version, find_links_dir, github_api_bearer_token
     )
+
+
+PYTHON_IDS = {
+    # N.B.: These values must match the lift TOML interpreter ids.
+    "cp38": "cpython38",
+    "cp39": "cpython39",
+    "cp310": "cpython310",
+    "cp311": "cpython311",
+}
+
+
+def determine_pex_name_and_python_id(
+    ptex: Ptex, version: Version, github_api_bearer_token: str | None = None
+) -> tuple[str, str]:
+    uname = os.uname()
+    platform = f"{uname.sysname.lower()}_{uname.machine.lower()}"
+    res = get_release_asset_and_python_id(ptex, version, platform, github_api_bearer_token)
+    if not res:
+        # If there's only one dot in version, specifically suggest adding the `.patch`
+        suggestion = (
+            "Pants version format not recognized. Please add `.<patch_version>` to the end of the version. "
+            "For example: `2.18` -> `2.18.0`.\n\n"
+            if version.base_version.count(".") < 2
+            else ""
+        )
+        fatal(
+            f"Unknown Pants release: {version}.\n\n{suggestion}"
+            f"Check to see if the URL is reachable ( {GITHUB_API_BASE_URL} ) and if"
+            f" an asset exists within the release for the platform {platform}."
+            " If the asset doesn't exist it may be that this platform isn't yet supported."
+            " If that's the case, please reach out on Slack: https://www.pantsbuild.org/community/getting-help#slack"
+            " or file an issue on GitHub: https://github.com/pantsbuild/pants/issues/new/choose.\n\n"
+        )
+    pex_name, python = res
+    if python not in PYTHON_IDS:
+        fatal(f"This version of scie-pants does not support {python!r}.")
+    return pex_name, PYTHON_IDS[python]
+
+
+GITHUB_API_BASE_URL = "https://api.github.com/repos/pantsbuild/pants"
+
+
+def fetch_github_api(ptex: Ptex, path: str, github_api_bearer_token: str | None = None) -> Any:
+    headers = (
+        {"Authorization": f"Bearer {github_api_bearer_token}"} if github_api_bearer_token else {}
+    )
+    return ptex.fetch_json(f"{GITHUB_API_BASE_URL}/{path}", **headers)
+
+
+def get_release_asset_and_python_id(
```

Comment:
Instead of using the API, what if we just took shots in the dark?

Since this is during Pants bootstrapping, the timing of the "failed" requests wouldn't matter much.

So, e.g. just try to see if we can download the asset with Py312, otherwise try again with py 3.11, and so on (or whatever the Python versions would be).

Reply:
That would work, of course. Do we need to? (i.e. is the api rate limit an actual obstacle here?)

There seems to be response headers we could examine to learn more, if we want to investigate further: https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#checking-the-status-of-your-rate-limit

Reply:
It just feels disingenuous to eat these on the users behalf without them knowing, since there are other ways.

Reply:
Actually, wait.

We'd never release a newer version of pants that uses a lower Python than the previous...

And when we bump the Python version, we're going to bump scie-pants (to add support). That means:
- either we know the Python version given a pants version. It's in the known set, or it isn't and we haven't bumped python yet
- or we bumped python and scie-pants doesn't know it

Meaning that the code should:
- use the specific Python version if it recognizes (e.g. 2.16 uses py39, 2.20 uses py39)
- use the latest Python version (if scie-pants of a particular version is running, and the last hardcoded version mapping says that Pants 2.25 uses py311, try py311 for pants version 2.29. either it _does_ use it, then yay, or it doesn't and scie-pants should be upgraded)

Does that make sense?

Reply:
> It just feels disingenuous to eat these on the users behalf without them knowing, since there are other ways.

That is a fair point.

> And when we bump the Python version, we're going to bump scie-pants (to add support).

Not sure I agree.. we can update scie-pants in advance to add the Python version not knowing at which time Pants will start using it.

I would prefer to have this de-coupled. We ship scie-pants with support for a number of Pythons.. and then it picks the correct one to use based on which version is needed for a particular version of Pants, which is not hard coded in scie-pants. Of course, historic versions could be hard coded to save on calls/time etc, and is likely the normal case most of the time, but it feels awkward to require a scie-pants release+upgrade whenever Pants bumps it's Python version.

Reply:
I see.

Yeah OK, maybe we combine the best of both worlds. Try-if-unknown, but any time we bump Python for Pants we also update scie-pants (so it is therefore not unknown). That way we stop firing blind, but also don't _require_ people upgrade.

Reply:
> Yeah OK, maybe we combine the best of both worlds

I like this idea.

Reply:
Yea, I'll rework this PR to break out tangential/unrelated changes per @huonw suggestion and fix the logic for determining Python version per @thejcannon suggestion.

Thanks for all reviews/comments so far :)

Reply:
I opted for building the expected URLs and doing a `HEAD` request to verify the Pants PEX exists as a way to determine what is available. We make an educated guess as to which Python version to use, failing that we try with all versions we support using.

i.e. There are no more Github API calls.